hypothesis-python/src/hypothesis/internal/conjecture/choice.py

```diff
@@ -456,13 +459,17 @@ def choice_permitted(choice: ChoiceT, kwargs: ChoiceKwargsT) -> bool:
         raise NotImplementedError(f"unhandled type {type(choice)} with value {choice}")
 
 
-def choices_key(choices: Sequence[ChoiceT]) -> tuple[ChoiceT, ...]:
+def choices_key(choices: Sequence[ChoiceT]) -> tuple[ChoiceKeyT, ...]:
     return tuple(choice_key(choice) for choice in choices)
 
 
-def choice_key(choice: ChoiceT) -> ChoiceT:
-    if type(choice) is float:
-        return float_to_int(choice)
+def choice_key(choice: ChoiceT) -> ChoiceKeyT:
+    if isinstance(choice, float):
+        # distinguish -0.0/0.0, signaling/nonsignaling nans, etc.
+        return ("float", float_to_int(choice))
```

Comment:
adding a "float" key here because otherwise `choice_key(f)` collides with `choice_key(float_to_int(f))`!

Reply:
actually, I just added this as a source comment.